### PR TITLE
Stabilize termination_trait, split out termination_trait_test

### PR DESCRIPTION
**Pull request by tmandry**
For #48453.

First time contribution, so I'd really appreciate any feedback on how this PR can be better.

Not sure exactly what kind of documentation update is needed. If there is no PR to update the reference, I can try doing that this week as I have time.

### CODE REVIEW

**Comment by tmandry** [comment]
cc @nikomatsakis

**Comment by nikomatsakis** [comment]
@tmandry 

> First time contribution, so I'd really appreciate any feedback on how this PR can be better.

Woohoo, welcome! <3

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-fn main() -> i32 { //~ ERROR main function has wrong type [E0580]
+fn main() -> i32 {
+//~^ ERROR the trait bound `i32: std::process::Termination` is not satisfied [E0277]
```
Comment (nikomatsakis) [inline src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: Hmm, this is not your fault, but this error message seems less clear than before.

@estebank -- can we improve this readily enough with a `#[rustc_on_unimplemented]` annotation perhaps?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great! My one concern is whether we can improve that error message in the case of a bad return type for `main`. I fear it is going to be confuse people as is.

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs (hunk 1 above)
Comment (estebank) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: Yes, you could annotate `std::process::Termination` using `#[rustc_on_unimplemented]` to replace the error message to something appropriate. The only limitation is not knowing wether the requirement came from a return type or anywhere else, so the message will have to be more generic than it is now.

Comment (tmandry) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: It looks like [there is already a `#[rustc_on_unimplemented]` annotation](https://github.com/rust-lang/rust/blob/e5a55e74405dedf8bc0744300a8c506eea94bc18/src/libstd/process.rs#L1445), but it is not used in the error message for some reason.

Comment (tmandry) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: Nevermind, once I cleaned my rustc build the note showed up. But since it's a label, not the primary error message, I had to use the `error-message:` directive to match against it.

**Comment by tmandry** [comment]
I updated the tests to reflect the error label that's already there from `rustc_on_unimplemented`. Here's what the full output looks like:

```
error[E0277]: the trait bound `i32: std::process::Termination` is not satisfied
  --> src/bin/termination-trait-main-i32.rs:11:18
   |
11 |   fn main() -> i32 {
   |  __________________^
12 | | //~^ ERROR the trait bound `i32: std::process::Termination` is not satisfied [E0277]
13 | |     0
14 | | }
   | |_^ `main` can only return types that implement std::process::Termination, not `i32`
   |
   = help: the trait `std::process::Termination` is not implemented for `i32`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

**Comment by nikomatsakis** [comment]
@tmandry ok, so, that is looking better! the main problem is that the span we are giving for the error (i.e., the region in the source code that we highlight) is the **function body**, which makes it cross multiple lines and makes my eyes glaze over (I imagine others have the same reaction). Let me see if I can give any tips about how to get a better span.

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -1106,25 +1106,22 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
...
+    // Check that the main return type implements the termination trait.
+    if let Some(term_id) = fcx.tcx.lang_items().termination() {
+        if let Some((id, _)) = *fcx.tcx.sess.entry_fn.borrow() {
+            if id == fn_id {
+                match fcx.sess().entry_type.get() {
+                    Some(config::EntryMain) => {
+                        let substs = fcx.tcx.mk_substs(iter::once(Kind::from(ret_ty)));
+                        let trait_ref = ty::TraitRef::new(term_id, substs);
+                        let cause = traits::ObligationCause::new(
+                            span, fn_id, ObligationCauseCode::MainFunctionType);
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: the span is specified here; if you look up to line 1028 you can see it comes from the body:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc_typeck/check/mod.rs#L1028

In general, to get a good span, you want to look to [the HIR](https://rust-lang-nursery.github.io/rustc-guide/hir.html), which corresponds pretty closely to the input program. In this case, we want to look at the argument `decl`:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc_typeck/check/mod.rs#L1001

The `FnDecl` struct is declared here:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc/hir/mod.rs#L1718-L1727

We want to get the span from the return type, for which there is a convenient method:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc/hir/mod.rs#L1815-L1820

So basically the span should be something like:

```rust
let return_ty_span = decl.output.span();
```

and use that instead of `span`.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs (hunk 1 above)
Comment (estebank) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: For what it's worth, you have further control on the `rustc_on_unimplemented` output by [being able to set a separate message and label](https://github.com/rust-lang/rust/pull/47613/files#diff-6f956da589a7eec6332ba1540c629560R533), as well as customize either one for specific types.

**Comment by tmandry** [comment]
Thanks for your help! I agree, the error message is much better this way. Updated and converted one of the tests to a UI test.

**Comment by nikomatsakis** [comment]
@tmandry 🏂

**Comment by nikomatsakis** [comment]
cc @rust-lang/docs -- I want to r+ this PR! It stabilizes `fn main() -> T { }` where `T: Termination` (though not the termination trait itself). What kind of documentation is needed and where?

**Comment by nikomatsakis** [comment]
[According to @aturon](https://gitter.im/rust-lang/WG-epoch?at=5ab168ba012ff2bf681db657), for edition-related docs we're going to handle them out of band. Therefore, I'm allowed to r+.

@bors r+

**Comment by nikomatsakis** [comment]
@tmandry nice work

**Review by nikomatsakis — APPROVED** [review]

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -551,6 +551,8 @@ declare_features! (
     (accepted, match_beginning_vert, "1.25.0", Some(44101), None),
     // Nested groups in `use` (RFC 2128)
     (accepted, use_nested_groups, "1.25.0", Some(44494), None),
+    // Termination trait in main (RFC 1937)
+    (accepted, termination_trait, "1.25.0", Some(43301), None),
```
Comment (scottmcm) [inline src/libsyntax/feature_gate.rs]: The build for `accepted` is when it was accepted, right?  Which I think will now be 1.26.

Also, you could consider making this feature be `termination_trait_main`, which would let all the remaining checks (and attributes in the minimal nightly code) stay unchanged.

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs [hunk 4]
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-fn main() -> i32 { //~ ERROR main function has wrong type [E0580]
+// error-pattern:`main` can only return types that implement std::process::Termination, not `i32`
```
Comment (scottmcm) [inline src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: It would be nice if this suggested some types that do work, like `()` and `!`.  (Especially since `Termination` is unstable, which makes these errors more confusing, based on reaction to `Carrier` errors.)

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (tmandry) [reply src/libsyntax/feature_gate.rs]: I wasn't sure about this. The tracking issue (#48453) has milestone 1.25.

re: `termination_trait_main`, it makes a little more sense since the trait itself is still not stable. However, it makes just as much sense to me to keep the trait for tests as it is now in the PR. Then we'd have `termination_trait_main`, `termination_trait_lib`, and `termination_trait_test`.

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs (hunk 1 above)
Comment (estebank) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: > since it's a label, not the primary error message, I had to use the error-message: directive to match against it.

You didn't need to use `error-message`, you could use 
```
//~^ NOTE `main` can only return types that implement std::process::Termination, not `i32`
```
instead for labels.

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs (hunk 4 above)
Comment (tmandry) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: I think that's a good idea. In particular, `Result<(), E> where E: Debug` helps me understand the motivation better. Is there any guidance for how long error messages can be?

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (tmandry) [reply src/libsyntax/feature_gate.rs]: After re-reading the [Forge guide](https://forge.rust-lang.org/stabilization-guide.html), it should definitely be 1.26. Updated.

For the feature name, I see where you're coming from more now: pull out the thing being stabilized, and leave the unstable stuff in the "main feature." I have no strong preference here at all, and went with the [initial recommendation](https://github.com/rust-lang/rust/issues/48453#issuecomment-372323844). If two people agree on what it should be, or one person feels more strongly and no one objects, I'll change it to that!

File: src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs (hunk 4 above)
Comment (tmandry) [reply src/test/compile-fail/rfc-1937-termination-trait/termination-trait-main-i32.rs]: I feel like fitting the `Result` suggestion into an error message is going to make this note too long no matter what, though I'm open to suggestions. So we are left with `()` and `!`.

Only question is, will a new Rust user from C who writes `fn main() -> i32` have any idea what `!` is, and if not, how would they Google it? `()` and `!` are both somewhat cryptic type names, and it seems like a bad experience for someone to come across error message with both of these types potentially very early on in their Rust language usage. That said, I can see how this would be helpful for people from other perspectives, and I'm sure I'm rehashing some argument from the original RFC. Perhaps we can use `never`.

For now, I've only included `()` as an example, with the expectation that someone who really wants to know what's going on can google `std::process::Termination` and read about other types that implement it. (I tried it, the first result gives all implementations!) I'd like to know what other people think about this as well.

**Comment by tmandry** [comment]
Updated to the right version (1.26) and added `()` as an example type to the error message.

I felt that adding other examples would either make the error message too long or too cryptic; see the [diff comments](https://github.com/rust-lang/rust/pull/49162#discussion_r175987283).

**Comment by nikomatsakis** [comment]
Hmm, so it occurs to me that since we are using `ObligationCauseCode::MainFunctionType` here, we actually have quite a lot of leeway to customize this however we want. In particular, I agree that we could use the "extended error" to give more details.

The idea would be to "intercept" trait error reporting when we see this cause code and do something completely different, rather like what happens here:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/traits/error_reporting.rs#L548-L559

Then we could arrange something like

```
error[E0580]: main does not have a valid return type
  --> $DIR/termination-trait-main-wrong-type.rs:11:14
   |
LL | fn main() -> char { //~ ERROR
   |              ^^^^ `main` can only return types that implement `std::process::Termination`
   |
   = help: consider using `()`, or a `Result`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0580`.
```

Moreover, we should ensure that the trait docs for `Termination` give examples and good guidance.

Thouhts @tmandry, @scottmcm, or others?

**Comment by tmandry** [comment]
@nikomatsakis I like it! I've pushed an example of doing this while keeping the E0277 message it has currently, since I'm not sure how to go from here to `librustc::infer::error_reporting` where E0580 is used (or vice versa). So currently it looks like 

```
error[E0277]: the trait bound `i32: std::process::Termination` is not satisfied
  --> src/bin/termination-trait-main-i32.rs:11:14
   |
11 | fn main() -> i32 {
   |              ^^^ `main` can only return types that implement std::process::Termination, not `i32`
   |
   = help: consider using `()`, or a `Result`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

File: src/libstd/process.rs [hunk 5]
```diff
@@ -1443,7 +1443,7 @@ pub fn id() -> u32 {
 #[cfg_attr(not(test), lang = "termination")]
 #[unstable(feature = "termination_trait_lib", issue = "43301")]
 #[rustc_on_unimplemented =
-  "`main` can only return types like `()` that implement {Termination}, not `{Self}`"]
+  "`main` can only return types that implement {Termination}, not `{Self}`"]
```
Comment (estebank) [inline src/libstd/process.rs]: You could change this to:

```
#[rustc_on_unimplemented(
  message="`main` has invalid return type `{Self}`",
  label="`main` can only return types that implement {Termination}")]
```

So that the output is

```
error[E0277]: `main` has invalid return type `i32`
  --> src/bin/termination-trait-main-i32.rs:11:14
   |
11 | fn main() -> i32 {
   |              ^^^ `main` can only return types that implement `std::process::Termination`
   |
   = help: consider using `()`, or a `Result`
```

Comment (tmandry) [reply src/libstd/process.rs]: That works. I wasn't sure how important going back to the original error code was.

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (scottmcm) [reply src/libsyntax/feature_gate.rs]: Good by me!  (I try to say "consider" for things I'm not that concerned about.)

**Comment by scottmcm** [comment]
That looks like a good direction, @nikomatsakis.

One thought: is there a canonical "I don't really care I just want to use `?`" `Result` type we could point people to?  Maybe `-> Result<(), Box<Debug>>` or something?  That's always doable later, though, and might be a libs domain to have a type alias like `MainResult` or something to make that easy for the simple cases...

**Comment by tmandry** [comment]
Thanks everyone for being so patient! Updated the message. It's way more helpful now!

```
error[E0277]: `main` has invalid return type `i32`
  --> src/bin/termination-trait-main-i32.rs:11:14
   |
11 | fn main() -> i32 {
   |              ^^^ `main` can only return types that implement std::process::Termination
   |
   = help: consider using `()`, or a `Result`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

**Comment by nikomatsakis** [comment]
Looks great.

**Comment by nikomatsakis** [comment]
So, it does look great, but I did just want to make two comments:

It does seem like it'd be marginally better to use the `E0580` code, so that we can have a `--explain` output that is tailored to this case. I can help @tmandry with the answer to this question:

> since I'm not sure how to go from here to librustc::infer::error_reporting where E0580 is used (or vice versa)

in one second.

Regarding what @scottmcm wrote:

> One thought: is there a canonical "I don't really care I just want to use ?" Result type we could point people to?

Interesting point. I don't think though that there is :( -- I wish that `Result<_, Box<Error>>` would work but for Reasons I don't think it will really will, and I don't think that `Box<dyn Debug>` would work because there needed `Into` clauses don't exist.

**Comment by nikomatsakis** [comment]
@bors r+

Ugh. OK, digging into the code I see what @tmandry was tempted to leave it the way they did. =) It's a bit annoying to use E0580. I'm sort of inclined to land this now, and leave that for possible follow-up.  So r+ again.

One interesting thing is that E0580 is also issued for things unrelated to the return type -- e.g., supplying arguments, or making main `unsafe` or the wrong ABI. So in a sense maybe using E0580 isn't the best option. If we *did* want to do it, we would create a refactored helper function somewhere that both the error-reporting modules can call into. But this code is in dire need of some refactoring.

Anyway, thanks @tmandry for all the effort you've put into this.

**Comment by estebank** [comment]
@nikomatsakis I just realized that it _could_ make sense to allow `rustc_on_unimplemented` to set an error code but I'd be worried about over/mis-use...

**Comment by scottmcm** [comment]
Filed https://github.com/rust-lang/rfcs/issues/2367 to track coming up with a good return type to suggest.
